packages/@aws-cdk/core/lib/app.ts

```diff
@@ -93,6 +96,18 @@ export class App extends Stage {
     return APP_SYMBOL in obj;
   }
 
+  /**
+   * Add a custom synthesis for the given construct
+   *
+   * When the construct is being synthesized, this allows it to add additional items
+   * into the Cloud Assembly output.
+   *
+   * This feature is intended for library authors; CDK users should almost never need to use this.
+   */
+  public static addCustomSynthesis(construct: Construct, synthesis: ICustomSynthesis): void {
```

Comment:
This API design allows 'any' construct to modify the synthesis output of any other construct. Possibility for abuse here by construct libraries and inject side-effect behaviour that's hard to account for.

Maybe we could add a method to `Resource` instead, or a magic method?

Neverthless, my preference is for none of these but instead to generate the buildspec in the constructor in combination with `Lazy`.

Reply:
> This API design allows 'any' construct to modify the synthesis output of any other construct. Possibility for abuse here by construct libraries and inject side-effect behaviour that's hard to account for.

I'm not sure how this is different if we do it at `app.synth()` time or in the constructor? 

Fact of the matter is you're running code on your machine that could use `fs.writeFileSync()` or `subprocess.spawn()` to do anything to any file on your machine. That capability is already there; at least if we provide a promoted API for it, we could build in checks and gates.

> Neverthless, my preference is for none of these but instead to generate the buildspec in the constructor in combination with Lazy.

I'm not seeing how that helps. I think you are proposing to do `fs.writeFileSync()` in the constructor? How will a `Lazy` help?